Source/Particles/Deposition/CurrentDeposition.H

```diff
@@ -0,0 +1,223 @@
+
+// Compute shape factor for order 1.
+template <>
+AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
+int compute_shape_factor <1> (Real* const sx, Real xmid){
+    int j = (int) xmid;
+    Real xint = xmid-j;
+    sx[0] = 1.0 - xint;
+    sx[1] = xint;
+    return j;
+}
+
+// Compute shape factor for order 2.
+template <>
+AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
+int compute_shape_factor <2> (Real* const sx, Real xmid){
+    int j = (int) (xmid+0.5);
+    Real xint = xmid-j;
+    sx[0] = 0.5*(0.5-xint)*(0.5-xint);
+    sx[1] = 0.75-xint*xint;
+    sx[2] = 0.5*(0.5+xint)*(0.5+xint);
+    return j;
+}
+
+// Compute shape factor for order 3.
+template <>
+AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
+int compute_shape_factor <3> (Real* const sx, Real xmid){
+    int j = (int) xmid;
+    Real xint = xmid-j;
+    sx[0] = 1.0/6.0*(1.0-xint)*(1.0-xint)*(1.0-xint);
+    sx[1] = 2.0/3.0-xint*xint*(1-xint/2.0);
+    sx[2] = 2.0/3.0-(1-xint)*(1-xint)*(1.0-0.5*(1-xint));
+    sx[3] = 1.0/6.0*xint*xint*xint;
+    return j;
+}
+
+/* \brief Current Deposition for thread thread_num
+ * /param xp, yp, zp   : Pointer to arrays of particle positions.
+ * \param wp           : Pointer to array of particle weights.
+ * \param uxp uyp uzp  : Pointer to arrays of particle momentum.
+ * \param jx_arr       : Array4 of current density, either full array or tile.
+ * \param jy_arr       : Array4 of current density, either full array or tile.
+ * \param jz_arr       : Array4 of current density, either full array or tile.
+ * \param offset       : Index of first particle for which current is deposited
+ * \param np_to_depose : Number of particles for which current is deposited.
+                        Particles [offset,offset+np_tp_depose] deposit current.
+ * \param dt           : Time step for particle level
+ * \param dx           : 3D cell size
+ * \param xyzmin       : Physical lower bounds of domain.
+ * \param lo           : Index lower bounds of domain.
+ * \param stagger_shift: 0 if nodal, 0.5 if staggered.
+ * /param q            : species charge.
+ */
+template <int depos_order>
+void doDepositionShapeN(const Real * const xp, const Real * const yp, const Real * const zp,
+                        const Real * const wp, const Real * const uxp,
+                        const Real * const uyp, const Real * const uzp,
+                        const amrex::Array4<amrex::Real>& jx_arr, 
+                        const amrex::Array4<amrex::Real>& jy_arr, 
+                        const amrex::Array4<amrex::Real>& jz_arr,
+                        const long offset, const long np_to_depose, 
+                        const amrex::Real dt, const std::array<amrex::Real,3>& dx,
+                        const std::array<Real, 3> xyzmin,
+                        const Dim3 lo,
+                        const amrex::Real stagger_shift, 
+                        const amrex::Real q)
+{
+    BL_PROFILE("PICSAR::CurrentDeposition");
```

Comment:
Not sure how useful the above line is, since this would reflect erroneous time, on GPU. (Unlike Fortran-OpenACC, the new C++ functions do not synchronize the device.)
@atmyers @WeiqunZhang Any suggestions?